src/perl/Permabit/BlockDevice/VDO/LVMVDO/ThinPool.pm

```diff
@@ -0,0 +1,176 @@
+#
+# $Id$
+##
+package Permabit::BlockDevice::VDO::LVMVDO::ThinPool;
+
+use strict;
+use warnings FATAL => qw(all);
+use Carp qw(confess);
+use English qw(-no_match_vars);
+use Log::Log4perl;
+use Permabit::Assertions qw(
+  assertDefined
+  assertNumArgs
+  assertType
+);
+use Permabit::Constants;
+use Permabit::ProcessUtils qw(delayFailures);
+use Permabit::Utils qw(makeFullPath);
+
+use base qw(
+  Permabit::BlockDevice::VDO::LVMVDO
+  Permabit::BlockDevice::LVM::ThinPool
+);
+
+my $log = Log::Log4perl->get_logger(__PACKAGE__);
+
+########################################################################
+# @inherit
+##
+sub setup {
+  my ($self) = assertNumArgs(1, @_);
+
+  # Get the location of the vdoformat binary for use by LVM.
+  $self->{vdoformat} = $self->getMachine()->findNamedExecutable("vdoformat");
+
+  # This will create a thin pool that has vdo set up to be the data device
+  my $config = $self->makeLVMConfigString();
+  my $totalSize = $self->getTotalSize($self->{physicalSize});
+  $self->{volumeGroup}->createThinPool($self->{deviceName},
+                                       $totalSize,
+                                       $config,
+                                       1);
+  $self->Permabit::BlockDevice::VDO::setup();
+}
+
+########################################################################
+# @inherit
+##
+sub growLogical {
+  my ($self, $logicalSize) = assertNumArgs(2, @_);
+
+  # resize the vdo logical volume
+  my $name = "$self->{deviceName}";
+
+  my $newSize = $self->{volumeGroup}->alignToExtent($logicalSize);
+  $self->{volumeGroup}->resizeLogicalVolume($name, $newSize);
+  $self->{logicalSize} = $newSize;
+}
+
+########################################################################
+# @inherit
+##
+sub growPhysical {
+  my ($self, $physicalSize) = assertNumArgs(2, @_);
+
+  # resize the vdo pool
+  my $name = "$self->{deviceName}pool";
+
+  my $newSize = $self->{metadataSize} + $physicalSize;
+  $newSize = $self->{volumeGroup}->alignToExtent($newSize);
+  $self->{volumeGroup}->resizeLogicalVolume($name, $newSize);
+  $self->{physicalSize} = $newSize - $self->{metadataSize};
+}
+
+########################################################################
+# Rename a VDO device. Currently LVM managed devices are the only
+# ones that support rename.
+#
+# @param newName The new name for the VDO device.
+##
+sub renameVDO {
+  my ($self, $newName) = assertNumArgs(2, @_);
+  my $name = "$self->{deviceName}";
+  $self->{volumeGroup}->renameLogicalVolume($name, $newName);
+  $self->setDeviceName($newName);
+}
+
+########################################################################
+# @inherit
+##
+sub disableWritableStorage {
+  my ($self) = assertNumArgs(1, @_);
+  my $deviceName = "$self->{deviceName}pool_vdata";
+  my $vgName   = $self->{volumeGroup}->getName();
+  my $fullName = "$vgName-$deviceName";
+
+  $self->runOnHost("sudo dmsetup remove $fullName");
+}
+
```

Comment:
The four readableStorage/writeableStorage methods are almost identical in the two implementations; the only difference is that Managed.pm calls the superclass method, while ThinPool doesn't. But looking at it, the superclass methods only set (or unset) a flag, so that tests can clean up properly if they get interrupted while in one of the disabled states.

So I think both versions should call the superclass method. From that, it also follows that these methods should migrate up into LVMVDO.pm.